### ISSUE TITLE
Filename* parameter not supported in Content-Disposition header

### ISSUE THREAD

**Issue by meetug** [SpiderLabs/ModSecurity#1825]
I tried the following request with the latest ModSec code with nginx/1.13.5 and CRS 3.0. It looks like the multipart parser does not support extension notation (RFC 5987) for filename parameter. I am getting the following error:

Error:
-----

ModSecurity: Access denied with code 400 (phase 2). Matched "Operator `Eq' with parameter `0' against variable `REQBODY_ERROR' (Value: `1' ) [file "modsecurity.conf"] [line "51"] [id "200002"] [rev ""] [msg "Failed to parse request body."] [data "Multipart parsing error: Multipart: Invalid Content-Disposition header (-14): form-data; name=document; filename=03CB16644685216C1E818D1369BB5D4B.png; filename*=utf-8''03CB16644685216C1E818D1369BB5D4B.png""] [severity "2"] [ver ""] [maturity "0"] [accuracy "0"] [hostname "167.220.24.83"] [uri "/"] [unique_id "153022593947.636509"] [ref "v247,1"]

Request:
---------
POST / HTTP/1.1
User-Agent: Fiddler
Accept: */* 
Cache-Control: no-cache
Postman-Token: 17ccb852-8244-40da-8cae-9464716ff393
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW

------WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name=document; filename=03CB16644685216C1E818D1369BB5D4B.png; filename*=utf-8''03CB16644685216C1E818D1369BB5D4B.png"
Content-Type: text/plain

------WebKitFormBoundary7MA4YWxkTrZu0gW--

**Comment by meetug** [comment]
Any ideas for a workaround?

**Comment by meetug** [comment]
Would it be possible for someone to look into this issue?

**Comment by victorhora** [comment]
Indeed currently the multipart parser does not take extension notation into account for the filename parameter.

For some reason your request hasn't triggered the error here, but the following example request from RFC 5987 does trigger the error:

```
POST /test.html HTTP/1.1
Host: example.org
Content-Type: multipart/form-data;boundary=boundary
Content-Length: 186

--boundary
Content-Disposition: form-data; name="field1"

value1
--boundary
Content-Disposition: form-data; name="field2"; filename*=UTF-8''%e2%82%ac%20rates

value2
--boundary--
```


Tagging as missing feature for libModSecurity.

The simplest "solution" is disabling rule IDs [200002](https://github.com/SpiderLabs/ModSecurity/blob/v2/master/modsecurity.conf-recommended#L59-L60) and [200003](https://github.com/SpiderLabs/ModSecurity/blob/v2/master/modsecurity.conf-recommended#L67-L81) from modsecurity.conf-recommended.

Depending on your setup, you could maybe do that only for part of your website that uses the unsupported feature (e.g. _/file-upload/_ or _upload.php_ ) so you would still have these rules working for the rest of the website.

**Comment by davewichers** [comment]
I'm having this problem as well in ModSecurity 2.9.1 , and so would love to know when this get's fixed in the 2.9.x branch as well, as 3.x.

**Comment by victorhora** [comment]
Adding to the 3.0.4 release milestone.

**Comment by davewichers** [comment]
Are you going to backport enhancements like this to the 2.9.x branch?

**Comment by victorhora** [comment]
I don't think so @davewichers 

Development efforts towards the 2.9.x branch is currently at maintenance level, meaning that security issues and/or major bugs are addressed with higher priority, while new features or minor bug fixes are meant for v3.

But being an Open Source project, everyone is welcomed to send patches/pull requests ;)

**Comment by davewichers** [comment]
Hmmm. It's my understanding that v3 really isn't ready for production use with Apache yet, like 2.9.x is. Is that correct? If not, are there instructions on how to seamlessly/easily upgrade/port from 2.9.x to 3.0.x? I couldn't find anything like that. If it is still Alpha, then seems like fixes like this that reduce false positives should be backported if at all possible.

**Comment by victorhora** [comment]
@davewichers Yes the Apache-connector is in development stage at this moment.

Documentation for the connector and libModSecurity can be found on the links below:
https://github.com/SpiderLabs/ModSecurity-apache/blob/master/README.md
https://github.com/SpiderLabs/ModSecurity/blob/v3/master/README.md

If you have more specific issues/questions about the Apache-connector, please direct them in the project page: https://github.com/SpiderLabs/ModSecurity-apache

As mentioned this an OpenSource project and we continue to accept fixes/enhancements to the 2.9.x branch, but the main focus of the very small team that actually and actively develops ModSecurity is for the v3 branch.

If there's anyone from the community willing to help with the addition (backporting might be easier as the logic is usually the same) of this feature to v2 I'm sure the ModSecurity team will gladly provide support during the development 🙂

Thanks.

**Comment by martinhsv** [comment]
Support for filename* is in v3/master as of https://github.com/SpiderLabs/ModSecurity/commit/136db3e582261382b668cc75592be667307dd052
